File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/infrastructure/persistence/RefundMapper.java
package com.example.accountservice.payment.infrastructure.persistence;

import com.example.accountservice.payment.domain.Refund;

/**
 * A class dedicated to converting between Refund (pure domain) and RefundJpaEntity (JPA mapping).
 * Used only inside RefundRepositoryImpl.
 */
final class RefundMapper {

    private RefundMapper() {}

    static Refund toDomain(RefundJpaEntity entity) {
        return Refund.reconstitute(
                entity.getRefundId(),
                entity.getPaymentId(),
                entity.getAmount(),
                entity.getReason(),
                entity.getStatus(),
                entity.getDecisionNote(),
                entity.getReasonCategory(),
                entity.getCreatedAt());
    }

    static RefundJpaEntity toNewEntity(Refund refund) {
        return new RefundJpaEntity(
                null,
                refund.getRefundId(),
                refund.getPaymentId(),
                refund.getAmount(),
                refund.getReason(),
                refund.getStatus(),
                refund.getDecisionNote(),
                refund.getReasonCategory(),
                refund.getCreatedAt());
    }

    static RefundJpaEntity updateEntity(RefundJpaEntity entity, Refund refund) {
        entity.applyMutableState(
                refund.getStatus(), refund.getDecisionNote(), refund.getReasonCategory());
        return entity;
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/infrastructure/persistence/RefundMapper.kt
package com.example.accountservice.payment.infrastructure.persistence

import com.example.accountservice.payment.domain.Refund

/**
 * The object dedicated to converting between Refund (pure domain) and RefundJpaEntity (JPA mapping).
 * Used only inside RefundRepositoryImpl.
 */
internal object RefundMapper {
    fun toDomain(entity: RefundJpaEntity): Refund =
        Refund.reconstitute(
            refundId = entity.refundId,
            paymentId = entity.paymentId,
            amount = entity.amount,
            reason = entity.reason,
            status = entity.status,
            decisionNote = entity.decisionNote,
            createdAt = entity.createdAt,
        )

    /** Creates a new entity (no PK, an insert target) for a new Refund. */
    fun toNewEntity(refund: Refund): RefundJpaEntity =
        RefundJpaEntity(
            id = null,
            refundId = refund.refundId,
            paymentId = refund.paymentId,
            amount = refund.amount,
            reason = refund.reason,
            status = refund.status,
            decisionNote = refund.decisionNote,
            createdAt = refund.createdAt,
        )

    /** Reflects the domain Refund's latest state onto the existing entity (preserving the PK) — an update target (approve/reject/complete). */
    fun updateEntity(
        entity: RefundJpaEntity,
        refund: Refund,
    ): RefundJpaEntity {
        entity.status = refund.status
        entity.decisionNote = refund.decisionNote
        return entity
    }
}
